// @nativescript/camera/common.js
export function getAspectSafeDimensions(sourceWidth, sourceHeight, reqWidth, reqHeight) {
    let widthCoef = sourceWidth / reqWidth;
    let heightCoef = sourceHeight / reqHeight;
    let aspectCoef = widthCoef > heightCoef ? widthCoef : heightCoef;
    return {
        width: Math.floor(sourceWidth / aspectCoef),
        height: Math.floor(sourceHeight / aspectCoef),
    };
}
function mapStatus(result) {
    let status = Status.unknown;
    if (result && result.length > 1) {
        if (Object.keys(Status).findIndex((k) => k === result[0]) >= 0) {
            status = Status[result[0]];
        }
    }
    return status;
}
export function mapError(e) {
    return {
        Success: false,
        Details: Status.error,
        Error: e,
    };
}
export function mapCameraPermissionStatus(permission) {
    const status = mapStatus(permission);
    const result = {
        Success: status === Status.authorized,
        Details: status,
    };
    return result;
}
export function mapPhotoPermissionStatus(permission) {
    const status = mapStatus(permission);
    const result = {
        Success: status === Status.authorized || status === Status.limited,
        Details: status,
    };
    return result;
}
export function combineCamerPhotoPermissions(cameraPermissions, photoPermissions) {
    const result = {
        Success: cameraPermissions.Success && photoPermissions.Success,
        Details: {
            Camera: cameraPermissions,
            Photo: photoPermissions,
        },
    };
    return result;
}
export var Status;
(function (Status) {
    Status["authorized"] = "authorized";
    Status["denied"] = "denied";
    Status["limited"] = "limited";
    Status["restricted"] = "restricted";
    Status["undetermined"] = "undetermined";
    Status["never_ask_again"] = "never_ask_again";
    Status["unknown"] = "unknown";
    Status["error"] = "error";
})(Status || (Status = {}));
